h2/src/main/org/h2/command/Command.java

```diff
@@ -242,7 +251,7 @@ public ResultInterface executeQuery(int maxrows, boolean scrollable) {
     public ResultWithGeneratedKeys executeUpdate(Object generatedKeysRequest) {
         long start = 0;
         Database database = session.getDatabase();
-        Object sync = database.isMultiThreaded() ? (Object) session : (Object) database;
+        Object sync = database.getMvStore() != null ? null : database.isMultiThreaded() ? session : database;
```

Comment:
`Session` is usually locked before invocation of this method in `JdbcStatement.executeUpdateInternal()`, `executeInternal()`, `executeQuery()` etc. With such change there may be situations where `Session` is not locked at all, because transactional commands and internal usages of `SET` command have own code paths in `JdbcConnection`/`JdbcStatement` without synchronization. Is it safe? If such command will be executed at the same time as an normal command it may affect fields of the session in an unexpected way. For example, it can clear generated keys, override scope identities, `currentCommandStart` and `cancelAtNs`.

Reply:
based an @katzyn 's comment, would be good to sprinkle around some 

    assert Thread.currentThread().holdsLock(sync);

Reply:
It's not legal to pass `null` to `Thread.holdsLock()` (this method is static, BTW) and such expectation even with non-null `sync` will be wrong.

From my point of view we should always lock at least session, because not all callers lock it before.

Also there is an another problem that I didn't notice before. `Session.setCurrentCommand()` should be in the same synchronized block as `update()`. Otherwise concurrent usage of command-specific fields in session will not be prevented.

It looks like changes in `Command.executeQuery()` and `Command.executeUpdate()` should be mostly reverted. But probably it's OK to ignore `MULTI_THREADED` in MVStore mode and lock the session even when `MULTI_THREAD` is not set.

Reply:
"Session is **usually** locked before invocation of this method" make me nervous :smile:. Ideally it should be **always** locked or not locked at all. I can't find example of " have own code paths in JdbcConnection/JdbcStatement without synchronization", you where talking about, but just to be on a safe side...
My concern was mostly about locking a database, so let's always sync it on session for MVStore case.

Reply:
It's locked when command is an application-provided SQL script from `Statement` or `PreparedStatement`, but many other methods do not lock session for unknown for me reason. For example, widely used `JdbcConnection.commit()` and `JdbcConnection.rollback()` do not lock the session. Methods that change some settings with `SET` command do not lock the session. So either all these methods should be reviewed and changed or at least session lock should be performed by `Command.executeQuery()` and `Command.executeUpdate()`.

And as I already wrote `Session.setCurrentCommand()` should be moved back inside the area where section is definitely locked. `Session` has own data that should not be used by different commands concurrently.